nixos/lib/test-driver/test_driver/machine.py

```diff
@@ -855,17 +855,30 @@ def screen_matches(last: bool) -> bool:
         with self.nested(f"waiting for {regex} to appear on screen"):
             retry(screen_matches)
 
-    def wait_for_console_text(self, regex: str) -> None:
+    def wait_for_console_text(self, regex: str, timeout: float | None = None) -> None:
+        """
+            Wait for the provided regex to appear on console.
+            For each reads,
+
+            If timeout is None, timeout is infinite.
+
+            `timeout` is in seconds.
+        """
         with self.nested(f"waiting for {regex} to appear on console"):
             # Buffer the console output, this is needed
             # to match multiline regexes.
             console = io.StringIO()
+            start = time.time()
             while True:
```

Comment:
Assuming nothing blocks so each iteraction takes negligible time, we could change this loop to have a counter, and not have to measure time.  That would even (maybe?) let us reuse `retry()`.

Reply:
Added a variant that reuses `retry`.